### DISCUSSION TITLE
Case-insensitive search function support for Table

### DISCUSSION THREAD

**Discussion by lethach** [dotnetcore/BootstrapBlazor#2483] [Q&A] [answered]
Case-insensitive search function support for Table. Thanks

**Comment by ArgoZhang** [comment]
@lethach hi. The `Table` component has no business logic. what's mean `Case-insensitive`

**Reply by mozahr** [reply]
Steps to reproduce

On table definition add ShowSearch = true
Define a column as searchable. 
If this column contains text value "SIMBA" and you try to search for "simba" no result will be returned.
But when i export, the data for simba is there...

**Reply by ArgoZhang** [reply]
@mozahr If a repro-project is not provided, this problem will be solved forever. Because I need to see how your code runs. display data provided by the user

**Comment by densen2014** [comment]
I don’t understand where the problem in Case-insensitive

![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/a897f228-bd83-4c11-bf97-8424354f64eb)
![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/f24bb30d-c425-4408-ac8c-05a2d07167a2)
![image](https://github.com/dotnetcore/BootstrapBlazor/assets/8428709/81f07438-3d2f-450f-b442-c848701a7282)

**Reply by mozahr** [reply]
Hello, 
you can go to you exemple
https://www.blazor.zone/table/search
2nd table **Auto-generate search function table** 
<img width="649" alt="Search with H capital" src="https://github.com/dotnetcore/BootstrapBlazor/assets/42202221/6ea0462a-2570-42a7-abd9-727ca88ae1cf">
<img width="540" alt="Search with same capital" src="https://github.com/dotnetcore/BootstrapBlazor/assets/42202221/869d83ef-5385-4ffa-87e8-1e8a75aea9cf">


Thank you

**Reply by ArgoZhang** [reply]
@lethach @mozahr it fixed release version `8.3.2-beta03`

**Reply by mozahr** [reply]
Ok thanks ! Will wait for full release then.

**Reply by mozahr** [reply]
Hello,
Tested the beta, and get this error
` System.InvalidOperationException: The LINQ expression 'DbSet<Item>()
          .Where(i => i.CodeBarre != null && i.CodeBarre.Contains(
              value: "simba",
              comparisonType: OrdinalIgnoreCase) || i.Title != null && i.Title.Contains(
              value: "simba",
              comparisonType: OrdinalIgnoreCase) || i.NomFournisseur != null && i.NomFournisseur.Contains(
              value: "simba",
              comparisonType: OrdinalIgnoreCase))' could not be translated. Additional information: Translation of method 'string.Contains' failed. If this method can be mapped to your custom function, see https://go.microsoft.com/fwlink/?linkid=2132413 for more information.`


Basically it cannot translate to SQL query.

Thanks

**Comment by ArgoZhang** [comment]
link issue [#3401](https://github.com/dotnetcore/BootstrapBlazor/issues/3041?wt.mc_id=DT-MVP-5004174)

**Comment by ArgoZhang** [comment]
@mozahr please take a look at this.

https://learn.microsoft.com/en-us/ef/core/miscellaneous/collations-and-case-sensitivity

`case-sensitivity` dependent on database

later I will roll back this fix and remove the `StringComparison` parameter. the normal `string.Contains` will be translated to `InStr` SQL function.

**Reply by mozahr** [reply]
Hello 

Did not work actually ...
Dont think the issue is related to the collation or StringComparaison...

Here is more info 
if you check the result here,

<img width="802" alt="Search result Empty with pagination" src="https://github.com/dotnetcore/BootstrapBlazor/assets/42202221/432734b6-a2f2-46aa-a695-0e265d7757e4">

you can see that the pagination is showing, but the result shows 0

Also if i inspect the result from the code, we can see all the items are listed

<img width="591" alt="image" src="https://github.com/dotnetcore/BootstrapBlazor/assets/42202221/c8cf4862-fd01-4881-b7c8-f0394ed62bfe">

Same result when we use capital letter.


Here is the code itemService

```
 public async Task<QueryData<Item>> OnQueryAsync(QueryPageOptions options, bool IsPaginated = true)
 {
     var _db = await _dbFactory.CreateDbContextAsync();
     var user = await _userService.GetCurrentUser();


     var query = _db.Set<Item>().Include(c => c.SousFamille).Include(c => c.Marque).Include("Images").AsNoTracking();

     if (user.Roles.Contains("ChefDeRayon"))
     {
         var sousFamilles = await _db.AspNetUserSousFamilles.Where(c => c.AspNetUserId == user.Id).Select(c => c.CodeSousFamille).ToListAsync();
         query = query.Where(c => sousFamilles.Contains((int)c.CodeSousFamille));
     }

     query = query
         .Where(options.Searchs.GetFilterLambda<Item>(FilterLogic.Or), options.Searchs.Any())
         .Where(options.Filters.GetFilterLambda<Item>(), options.Filters.Any())
         .Sort(options.SortName!, options.SortOrder, !string.IsNullOrEmpty(options.SortName))
         .Count(out var count);



     if (IsPaginated)
         query = query.Page((options.PageIndex - 1) * options.PageItems, options.PageItems);


     var ret = new QueryData<Item>()
     {
         TotalCount = count,
         Items = query,
         IsSorted = true,
         IsFiltered = true
     };
     return await Task.FromResult(ret);
 }
```
 

code from item razor page
```
 private async Task<QueryData<Item>> OnQueryAsync(QueryPageOptions options)
 {

     currentOptions = options;
     var result =  await _itemService.OnQueryAsync(options);
     return result;
 }
```

Also when we use the column filter,

<img width="218" alt="image" src="https://github.com/dotnetcore/BootstrapBlazor/assets/42202221/7b3103e5-a788-4377-9bec-3c8e26457b41">

it works perfectly.

So in conclusion, it seems that when the data is passed to your table, pagination is well handled but somehow the data are not display.

Thank you..

**Reply by ArgoZhang** [reply, marked as answer]
```
var ret = new QueryData<Item>()
 {
     TotalCount = count,
     Items = query,
     IsSorted = true,
     IsFiltered = true
 };
```

missing code 
```
IsSearch = true
```

so why hasn't the code been provided? With the code, we know where the problem is.

**Reply by mozahr** [reply]
Unbeleivable.. 

Thanks so much.
